// @angular/service-worker/worker/bootstrap.js
import { NgSwCacheImpl, NgSwEvents, NgSwFetch } from './facade';
import { Driver } from './driver';
// The scope is the global object.
var scope = ((typeof self !== 'undefined') ? self : global);
function copyRequest(req) {
    var copy = {
        method: req.method,
        headers: req.headers,
        credentials: req.credentials,
        cache: req.cache,
        redirect: req.redirect,
        referrer: req.referrer,
    };
    if (req.mode.toString() !== 'navigate') {
        copy['mode'] = req.mode;
    }
    return copy;
}
var NgSwBrowserAdapter = (function () {
    function NgSwBrowserAdapter() {
    }
    NgSwBrowserAdapter.prototype.newRequest = function (req, init) {
        if (init && init instanceof Request) {
            init = copyRequest(init);
        }
        return new Request(req, init);
    };
    NgSwBrowserAdapter.prototype.newResponse = function (body) {
        return new Response(body);
    };
    return NgSwBrowserAdapter;
}());
export function bootstrapServiceWorker(options) {
    var manifestUrl = (options && options.manifestUrl) || '/ngsw-manifest.json';
    var plugins = (options && options.plugins) || [];
    var adapter = new NgSwBrowserAdapter();
    var cache = new NgSwCacheImpl(scope.caches, adapter);
    var events = new NgSwEvents(scope);
    var fetch = new NgSwFetch(scope, adapter);
    return new Driver(manifestUrl, plugins, scope, adapter, cache, events, fetch);
}
